src/spyglass/spikesorting/v1/recording.py

```diff
@@ -0,0 +1,876 @@
+
+        probe_type_by_channel = []
+        electrode_group_by_channel = []
+        for channel_id in channel_ids:
+            probe_type_by_channel.append(
+                (
+                    Electrode * Probe
+                    & {
+                        "nwb_file_name": nwb_file_name,
+                        "electrode_id": channel_id,
+                    }
+                ).fetch1("probe_type")
+            )
+            electrode_group_by_channel.append(
+                (
+                    Electrode
+                    & {
+                        "nwb_file_name": nwb_file_name,
+                        "electrode_id": channel_id,
+                    }
+                ).fetch1("electrode_group_name")
+            )
+        probe_type = np.unique(probe_type_by_channel)
+        filter_params = (
+            SpikeSortingPreprocessingParameters * SpikeSortingRecordingSelection
+            & key
+        ).fetch1("preproc_params")
+
+        # DO:
+        # - load NWB file as a spikeinterface Recording
+        # - slice the recording object in time and channels
+        # - apply referencing depending on the option chosen by the user
+        # - apply bandpass filter
+        # - set probe to recording
+        recording = se.read_nwb_recording(
+            nwb_file_abs_path, load_time_vector=True
+        )
+        all_timestamps = recording.get_times()
+
+        # TODO: make sure the following works for recordings that don't have explicit timestamps
+        valid_sort_times = self._get_sort_interval_valid_times(key)
+        valid_sort_times_indices = _consolidate_intervals(
+            valid_sort_times, all_timestamps
+        )
+
+        # slice in time; concatenate disjoint sort intervals
+        if len(valid_sort_times_indices) > 1:
+            recordings_list = []
+            timestamps = []
+            for interval_indices in valid_sort_times_indices:
+                recording_single = recording.frame_slice(
+                    start_frame=interval_indices[0],
+                    end_frame=interval_indices[1],
+                )
+                recordings_list.append(recording_single)
+                timestamps.append(
+                    all_timestamps[
+                        interval_indices[0] : interval_indices[1]
+                    ]
+                )
+            recording = si.concatenate_recordings(recordings_list)
+        else:
+            recording = recording.frame_slice(
+                start_frame=valid_sort_times_indices[0][0],
+                end_frame=valid_sort_times_indices[0][1],
+            )
+            timestamps = all_timestamps[
+                valid_sort_times_indices[0][0] : valid_sort_times_indices[0][1]
+            ]
+
+        # slice in channels; include ref channel in first slice, then exclude it in second slice
+        if ref_channel_id >= 0:
+            recording = recording.channel_slice(channel_ids=channel_ids)
```

Comment:
If `ref_channel` is not part of the sort group, then the recording obviously can't be referenced. So it seems like the error is expected. On the other hand, if you do correctly specify the `ref_channel` in your sort group, then you get back the non-ref channels after referencing (since the ref channel would just be zero). Is this not the correct behavior? What do you think it should be?

Reply:
I think this is an issue for something like tetrodes where you wank to reference the data on one tetrode against a reference on another.  Since it would be redundant to put the reference electrode in every group the previous version let you reference against one not in the group.  The `set_group_by_shank` function in the previous SortGroup table depended on this functionality

The previous version allowed it to work in the same way I suggested here. Previous version:
```
# include ref channel in first slice, then exclude it in second slice
if ref_channel_id >= 0:
    channel_ids_ref = np.append(channel_ids, ref_channel_id)
    recording = recording.channel_slice(channel_ids=channel_ids_ref)

    recording = si.preprocessing.common_reference(
        recording, reference="single", ref_channel_ids=ref_channel_id
    )
    recording = recording.channel_slice(channel_ids=channel_ids)
```

Reply:
I see, I made changes so that even if the ref channel is not one of the channels in the sort group, it is used to reference them.